### DISCUSSION TITLE
Clarification on the definition of stress/virial

### DISCUSSION THREAD

**Discussion by weilaigit** [mir-group/nequip#323] [Q&A] [answered]
I would appreciate it if the definition of stress/virial in NequIP could be clarified.  

1.  It seems that stress and virial are used interexchangely in NequIP.  My understanding is that the stress includes both the kinetic and potential (virial) part.  In the training, using stress and virial interexchangely is fine as only the virial in the training data is relevant (stress=virial in this case).  However, the stress is relevant in the NPT MD simulation so I want to confirm it is the stress, not virial, that was the output of StressForceOutput(https://github.com/mir-group/nequip/discussions/309), that was used in LAMMPS to run NPT simulations.
2. Stress/virial can be expressed in the unit of energy or pressure.  In LAMMPS, they are both in the unit of pressure.  However, in another discussion here about including virial in the loss function (https://github.com/mir-group/nequip/discussions/312), it was mentioned that virial was an extensive property.  I am a little confused.

**Comment by Linux-cpp-lisp** [comment, marked as answer]
Hi @weilaigit ,

Yes, I think there is a difference in terms here. In `nequip` "stress" and "virial" are both gradients of the **potential** energy, but the differ by a sign flip and a factor of the volume $V$ of the box:
  - "Virial" means $-\nabla_{\text{box}}E_\text{potential}$, a symmetric 3x3 tensor
  - "Stress" means $-\frac{1}{V} \times \text{virial} = \frac{1}{V} \nabla_{\text{box}}E_\text{potential}$

The sign convention on and between these two quantities is, to the best of my understanding, arbitrary and varies between people and DFT codes. See https://github.com/mir-group/nequip/blob/main/nequip/nn/_grad_output.py#L327 for more discussion and links on the convention.

> It seems that stress and virial are used interexchangely

No, but we support them in many of the same places:  `StressForceOutput` predicts both stress and virial, you can include stress or virial in your training data + loss function, and so on.

> that was used in LAMMPS to run NPT simulations.

See https://github.com/mir-group/pair_nequip/blob/stress/pair_nequip.cpp#L478.  We set the LAMMPS virial to our virial, and let LAMMPS handle the rest in terms of converting into pressure units, adding the kinetic term, etc.

> discussion here about including virial in the loss function

In `nequip` virial is in energy units and stress is in energy / length^3 units, **which is different than typical pressure units**. We call these the "self consistent" units in a few discussions, since they are derived from the base energy and length units. Training data must be provided in those units.

**Reply by weilaigit** [reply]
@Linux-cpp-lisp Thanks for the clarification.
